### PR TITLE
Token granting

### PR DESCRIPTION
**Pull request by thebevrishot**
This PR is not include EndToEnd test.

closes https://github.com/zcoinofficial/ztm/issues/82 https://github.com/zcoinofficial/ztm/issues/154

### CODE REVIEW

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+using Ztm.WebApi.Watchers.TransactionConfirmation;
+using Ztm.Zcoin.NBitcoin.Exodus;
+using Ztm.Zcoin.Rpc;
+using Transaction = Ztm.WebApi.Models.Transaction;
+
+namespace Ztm.WebApi.Controllers
+{
+    [Route("api")]
+    [ApiController]
+    public class TokenController : ControllerBase
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: I think we should make this controller as a dedicated to `issue-tokens` only (e.g. `IssueTokenController`).

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+using Ztm.WebApi.Callbacks;
+using Ztm.WebApi.Models;
+using Ztm.WebApi.Watchers.TransactionConfirmation;
+using Ztm.Zcoin.NBitcoin.Exodus;
+using Ztm.Zcoin.Rpc;
+using Transaction = Ztm.WebApi.Models.Transaction;
+
+namespace Ztm.WebApi.Controllers
+{
+    [Route("api")]
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: We should not prefix URL with `api`. So just `issue-tokens` is enough.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            }
+
+            this.factory = factory;
+            this.configuration = configuration;
+            this.watcher = watcher;
+            this.callbackRepository = callbackRepository;
+            this.ruleRepository = ruleRepository;
+        }
+
+        [HttpPost("issue-tokens")]
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: Once we updated as my previous comment we need to update this.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        public async Task<IActionResult> Issue([FromBody] Issuing issueing)
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: Same here. Just `PostAsync` is better.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        {
+            var zcoinConfig = this.configuration.GetZcoinSection();
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: We should cache the configuration in the constructor.

File: src/Ztm.WebApi/Controllers/TokenController.cs (hunk 4 above)
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: Rename `issueing` to `req` looks better.

Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: We should accept `CancellationToken` too.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+            using (var client = await this.factory.CreateRpcClientAsync(CancellationToken.None))
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: Once we accept `CancellationToken` we should allow this to cancel.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            {
+                var property = new Property(zcoinConfig.Property.Id, zcoinConfig.Property.Type);
+
+                var tx = await client.GrantPropertyAsync
+                (
+                    property,
+                    zcoinConfig.Property.Distributor.Address,
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: This is not correct. For token issuing we will always grant from `issuer` to `distributor`.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                    issueing.Destination,
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: Same here.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+                    issueing.Amount,
+                    issueing.Note,
+                    CancellationToken.None
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: Same as previous issue.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+                    (
+                        id,
+                        new CallbackResult("success", callbackResult),
+                        new CallbackResult("tokens-issuing-timeout", callbackResult),
+                        callback,
+                        CancellationToken.None
+                    );
+                }
+
+                return Ok(new Transaction{Tx = id, Fee = info.Fee});
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: Space error before `{`.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+                var callback = await this.AddCallbackAsync(CancellationToken.None);
+                if (callback != null)
+                {
+                    var callbackResult = new IssuingCallback{Tx=id};
+
+                    await this.AddRuleAsync
+                    (
+                        id,
+                        new CallbackResult("success", callbackResult),
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: Better to use constant for success status.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+            var callback = await this.callbackRepository.AddAsync(this.HttpContext.Connection.RemoteIpAddress, url, cancellationToken);
+            this.SetCallbackId(callback.Id);
+
+            return callback;
+        }
+
+        Task<Rule> AddRuleAsync(uint256 id, CallbackResult success, CallbackResult timeout, Callback callback, CancellationToken cancellationToken)
+        {
+            var config = this.configuration.GetCallbackSection();
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: Same as previous issue.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+                return null;
+            }
+
+            var callback = await this.callbackRepository.AddAsync(this.HttpContext.Connection.RemoteIpAddress, url, cancellationToken);
+            this.SetCallbackId(callback.Id);
+
+            return callback;
+        }
+
+        Task<Rule> AddRuleAsync(uint256 id, CallbackResult success, CallbackResult timeout, Callback callback, CancellationToken cancellationToken)
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: `AddRuleAsync` is quite confusing. IMO rename to `WatchTransactionAsync` looks better.

File: src/Ztm.WebApi/Controllers/TokenController.cs [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        {
+            var config = this.configuration.GetCallbackSection();
+            return this.watcher.AddTransactionAsync(
+                id, config.TransactionConfirmation.RequiredConfirmation, config.TransactionConfirmation.Timeout, callback, success, timeout, cancellationToken);
+        }
+    }
+
+    struct IssuingCallback
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Controllers/TokenController.cs]: We can remove this and use anonymous type instead.

Comment (thebevrishot) [reply src/Ztm.WebApi/Controllers/TokenController.cs]: good idea

File: src/Ztm.WebApi/Controllers/TokenController.cs (hunk 10 above)
Comment (thebevrishot) [reply src/Ztm.WebApi/Controllers/TokenController.cs]: Move to next line instead

File: src/Ztm.WebApi/Watchers/TransactionConfirmation/EntityRuleRepository.cs [hunk 15]
```diff
@@ -10,38 +10,51 @@
 using Ztm.Data.Entity.Contexts;
 using Ztm.WebApi.Callbacks;
 using EntityModel = Ztm.Data.Entity.Contexts.Main.TransactionConfirmationWatcherRule;
+using System.Text;
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Watchers/TransactionConfirmation/EntityRuleRepository.cs]: This one should come before type alias.

File: src/Ztm.WebApi/Watchers/TransactionConfirmation/EntityRuleRepository.cs [hunk 16]
```diff
@@ -10,38 +10,51 @@
...
+using System.IO;
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Watchers/TransactionConfirmation/EntityRuleRepository.cs]: Same here.

File: src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs [hunk 17]
```diff
@@ -172,51 +189,45 @@ public async Task StopAsync(CancellationToken cancellationToken)
...
 
-                timers.Add(rule.Id, timer);
+            timers.Add(rule.Id, timer);
 
-                try
-                {
-                    var remainingWaitingTime = await this.ruleRepository.GetRemainingWaitingTimeAsync(rule.Id, CancellationToken.None);
+            try
+            {
+                await Task.Delay(100);
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs]: Why we need to delay here?

File: src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs [hunk 18]
```diff
@@ -172,51 +189,45 @@ public async Task StopAsync(CancellationToken cancellationToken)
...
+        // Thread Unsafe
         async Task SetupTimerAsync(Rule rule)
         {
             var timer = new Timer();
 
-            this.timerLock.EnterWriteLock();
-            try
-            {
-                Dictionary<Guid, Timer> timers;
+            Dictionary<Guid, Timer> timers;
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs]: We should check if `timersSemaphore` already locked here.

File: src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs [hunk 19]
```diff
@@ -172,51 +189,45 @@ public async Task StopAsync(CancellationToken cancellationToken)
             }
             finally
             {
-                this.timerLock.ExitWriteLock();
+                this.timersSemaphore.Release();
             }
         }
 
+        // Thread Unsafe
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs]: Once we followed my previous idea we can remove this.

File: src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs [hunk 20]
```diff
@@ -231,57 +242,41 @@ void OnTimeout(object sender, TimerElapsedEventArgs e)
...
             });
         }
 
+        // Thread Unsafe
         void RemoveTimer(Rule rule)
         {
-            this.timerLock.EnterWriteLock();
-
-            try
+            if (!this.timers.TryGetValue(rule.TransactionHash, out var timers))
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs]: Same as previous issue.

File: src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs [hunk 21]
```diff
@@ -231,57 +242,41 @@ void OnTimeout(object sender, TimerElapsedEventArgs e)
                 finally
                 {
                     RemoveTimer(rule);
-                    this.timerLock.ExitWriteLock();
+                    this.timersSemaphore.Release();
                 }
             });
         }
 
+        // Thread Unsafe
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs]: Same here.

File: src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs [hunk 22]
```diff
@@ -231,57 +242,41 @@ void OnTimeout(object sender, TimerElapsedEventArgs e)
...
             }
         }
 
+        // Thread Unsafe
         async Task<bool> StopTimer(Rule rule)
         {
-            this.timerLock.EnterWriteLock();
-
-            try
+            if (this.timers.TryGetValue(rule.TransactionHash, out var txTimers) && txTimers.TryGetValue(rule.Id, out var timer))
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs]: Same here.

File: src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs [hunk 23]
```diff
@@ -231,57 +242,41 @@ void OnTimeout(object sender, TimerElapsedEventArgs e)
...
-            }
-            finally
+            if (timers.Count == 0)
             {
-                this.timerLock.ExitWriteLock();
+                this.timers.Remove(rule.TransactionHash);
             }
         }
 
+        // Thread Unsafe
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs]: Same here.

File: src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs [hunk 24]
```diff
@@ -378,32 +373,50 @@ async Task<IEnumerable<Rule>> ITransactionConfirmationWatcherHandler<Rule>.Creat
...
-            {
-                throw new ArgumentNullException(nameof(watches));
-            }
+            await this.timersSemaphore.WaitAsync();
 
-            foreach (var watch in watches) // lgtm [cs/linq/missed-where]
+            try
             {
-                if (await StopTimer(watch.Context))
+                if (watches == null)
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Watchers/TransactionConfirmation/TransactionConfirmationWatcher.cs]: Better to put this one outside critical section.

File: src/Ztm.Zcoin.NBitcoin/UInt256JsonConverter.cs [hunk 25]
```diff
@@ -0,0 +1,19 @@
+using System;
+using NBitcoin;
+using Newtonsoft.Json;
+
+namespace Ztm.Zcoin.NBitcoin
+{
+    public sealed class UInt256JsonConverter : JsonConverter<uint256>
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/UInt256JsonConverter.cs]: We should move this one to a separated project (e.g. `Ztm.Zcoin.NBitcoin.Json`) due to in required additional dependencies.

File: src/Ztm.Zcoin.NBitcoin/BitcoinAddressJsonConverter.cs [hunk 26]
```diff
@@ -0,0 +1,32 @@
+using System;
+using NBitcoin;
+using Newtonsoft.Json;
+
+namespace Ztm.Zcoin.NBitcoin
+{
+    public sealed class BitcoinAddressJsonConverter : JsonConverter<BitcoinAddress>
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.NBitcoin/BitcoinAddressJsonConverter.cs]: Same as previous issue.

File: src/Ztm.WebApi/Startup.cs [hunk 27]
```diff
@@ -108,6 +137,7 @@ void ConfigureMvc(MvcOptions options)
             // Custom Model Binders.
             options.ModelBinderProviders.Insert(0, new BitcoinAddressModelBinderProvider());
             options.ModelBinderProviders.Insert(0, new PropertyAmountModelBinderProvider());
+            options.ModelMetadataDetailsProviders.Add(new SuppressChildValidationMetadataProvider(typeof(BitcoinAddress)));
```
Comment (ultimaweapon) [inline src/Ztm.WebApi/Startup.cs]: Why do we need to suppress validation for `BitcoinAddress`?

Comment (thebevrishot) [reply src/Ztm.WebApi/Startup.cs]: This works around the infinity recursion issue after parse `BitcoinAddress`.

File: specifications.md [hunk 28]
```diff
@@ -33,28 +33,31 @@ Sample request to issue 1,000,000 tokens:
 ```json
 {
   "amount": 100000000000000,
```
[the remark this answers is not in the document]
Comment (thebevrishot) [reply specifications.md]: nice catch!!

File: src/Ztm.WebApi/Startup.cs (hunk 27 above)
Comment (ultimaweapon) [reply src/Ztm.WebApi/Startup.cs]: I see. Better to put `FIXME` here so people can remove it once we upgrade the framework.

File: src/Ztm.WebApi.Tests/Controllers/IssueTokenControllerTests.cs [hunk 29]
```diff
@@ -0,0 +1,297 @@
...
+            // Assert.
+            this.propertyManagementRpc.Verify();
+            this.rawTransactionRpc.Verify();
+
+            var okObjectResult = Assert.IsType<OkObjectResult>(result);
+            var returnedTx = okObjectResult.Value;
+            var properties = returnedTx.GetType().GetProperties();
+            var txVal = (uint256)properties.First(p => p.Name == "Tx").GetValue(returnedTx);
+
+            Assert.Equal(tx.GetHash(), txVal);
```
Comment (ultimaweapon) [inline src/Ztm.WebApi.Tests/Controllers/IssueTokenControllerTests.cs]: Better to use [FluentAssertion](https://fluentassertions.com/objectgraphs/) on this so we don't need to use reflection like this.

File: src/Ztm.WebApi.Tests/Controllers/IssueTokenControllerTests.cs [hunk 30]
```diff
@@ -0,0 +1,297 @@
...
+        }
+
+        [Fact]
+        public async Task PostAsync_WithValidArgumentAndUrlWasNotSet_ShouldSuccessButNotAddCallback()
+        {
+            // Arrange.
+            var amount = PropertyAmount.One;
+            var note = "Test Issuing";
+            var tx = NBitcoin.Transaction.Parse(TestTransaction.Raw1, ZcoinNetworks.Instance.Mainnet);
+            var fee = Money.Satoshis(500);
```
Comment (ultimaweapon) [inline src/Ztm.WebApi.Tests/Controllers/IssueTokenControllerTests.cs]: Seems like we don't need this.

File: src/Ztm.WebApi.Tests/Controllers/IssueTokenControllerTests.cs [hunk 31]
```diff
@@ -0,0 +1,297 @@
...
+        }
+
+        [Fact]
+        public async Task PostAsync_WithValidArgumentAndCallbackUrlIsSet_ShouldSuccess()
+        {
+            // Arrange.
+            var amount = PropertyAmount.One;
+            var note = "Test Issuing";
+            var tx = NBitcoin.Transaction.Parse(TestTransaction.Raw1, ZcoinNetworks.Instance.Mainnet);
+            var fee = Money.Satoshis(500);
```
Comment (ultimaweapon) [inline src/Ztm.WebApi.Tests/Controllers/IssueTokenControllerTests.cs]: Same here.

File: src/Ztm.WebApi.Tests/Controllers/IssueTokenControllerTests.cs [hunk 32]
```diff
@@ -0,0 +1,297 @@
...
+            // Assert.
+            this.propertyManagementRpc.Verify();
+            this.rawTransactionRpc.Verify();
+
+            var okObjectResult = Assert.IsType<OkObjectResult>(result);
+            var returnedTx = okObjectResult.Value;
+            var properties = returnedTx.GetType().GetProperties();
+            var txVal = (uint256)properties.First(p => p.Name == "Tx").GetValue(returnedTx);
+
+            Assert.Equal(tx.GetHash(), txVal);
```
Comment (ultimaweapon) [inline src/Ztm.WebApi.Tests/Controllers/IssueTokenControllerTests.cs]: Same as previous issue.

File: src/Ztm.WebApi/Startup.cs (hunk 27 above)
Comment (thebevrishot) [reply src/Ztm.WebApi/Startup.cs]: Good idea

**Pull request merged**